keccak256/src/circuit.rs

```diff
@@ -1,42 +1,43 @@
 use super::gates::{
-    absorb::{AbsorbConfig, ABSORB_NEXT_INPUTS},
-    iota_b13::IotaB13Config,
-    iota_b9::IotaB9Config,
-    pi::PiConfig,
-    rho::RhoConfig,
-    theta::ThetaConfig,
+    iota_b9::IotaB9Config, pi::pi_gate_permutation, rho::RhoConfig,
+    state_conversion::StateBaseConversion, tables::FromBase9TableConfig, theta::ThetaConfig,
     xi::XiConfig,
 };
 use crate::{
     arith_helpers::*,
-    common::{PERMUTATION, ROTATION_CONSTANTS, ROUND_CONSTANTS},
-    gates::rho_checks::RhoAdvices,
+    common::{NEXT_INPUTS_LANES, PERMUTATION, ROUND_CONSTANTS},
+    gates::tables::{Base13toBase9TableConfig, SpecialChunkTableConfig},
 };
 use crate::{gates::mixing::MixingConfig, keccak_arith::*};
 use halo2::{
-    circuit::Region,
+    circuit::{Cell, Layouter, Region},
     plonk::{Advice, Column, ConstraintSystem, Error, Selector},
+    poly::Rotation,
 };
 use itertools::Itertools;
-use num_bigint::BigUint;
-use pasta_curves::arithmetic::FieldExt;
-use std::{convert::TryInto, marker::PhantomData};
+use pairing::arithmetic::FieldExt;
+use std::convert::TryInto;
 
 #[derive(Clone, Debug)]
 pub struct KeccakFConfig<F: FieldExt> {
     theta_config: ThetaConfig<F>,
     rho_config: RhoConfig<F>,
-    pi_config: PiConfig<F>,
     xi_config: XiConfig<F>,
     iota_b9_config: IotaB9Config<F>,
+    base_conversion_config: StateBaseConversion<F>,
     mixing_config: MixingConfig<F>,
-    state: [Column<Advice>; 25],
-    _marker: PhantomData<F>,
+    pub state: [Column<Advice>; 25],
+    q_out: Selector,
+    pub _is_mixing_flag: Column<Advice>,
+    _base_conv_activator: Column<Advice>,
```

Comment:
Yes. 

`base_conv_activator` is always `1` since it's needed in order to activate the base conversion gadget in the loop of the keccak permutation. (Where Mixing Step is not called).

Instead, `is_mixing_flag` is just used on that last iteration when we perform the `Mixing`.

Reply:
I see, just saw the `TODO` to change `_base_conv_activator` to a `Selector`.